BFG closing a closed channel

### Confirmation

- [X] I have checked for similar issues.

### Describe the bug

BFG restarted on this error
```
panic: close of closed channel

goroutine 1277685 [running]:
github.com/hemilabs/heminetwork/hemi/electrumx.(*clientConn).Close(0xc001b6cba0)
	github.com/hemilabs/heminetwork/hemi/electrumx/conn.go:184 +0x10e
github.com/hemilabs/heminetwork/hemi/electrumx.(*clientConn).pinger(0xc001b6cba0)
	github.com/hemilabs/heminetwork/hemi/electrumx/conn.go:166 +0x152
created by github.com/hemilabs/heminetwork/hemi/electrumx.newClientConn in goroutine 1277683
	github.com/hemilabs/heminetwork/hemi/electrumx/conn.go:43 +0xd6
```

### Expected behaviour

We shouldn't be double closing channels.

### Environment

Version: <!-- Hemi Network v1.2.3 -->
Operating System: <!-- e.g. Windows 11, macOS 14.2.1, Ubuntu 24.04 -->
Architecture: <!-- amd64, arm64 -->


### Additional Information

The service was restarted upon failure and this error hasn't occurred since.  This can be low priority at the moment as it's rare and we can quickly recover.

--- a/hemi/electrumx/conn.go
+++ b/hemi/electrumx/conn.go
@@ -186,13 +186,14 @@ func (c *clientConn) Close() error {
 	if c.onClose != nil {
 		c.onClose(c)
 	}
-	if c.closeCh != nil {
-		close(c.closeCh)
-		c.closeCh = nil
-	}
 
 	if err := c.conn.Close(); err != nil {
 		return err
+	}
+
+	if c.closeCh != nil {
+		close(c.closeCh)
+		c.closeCh = nil
 	}
 
 	// Wait for pinger to exit.

--- a/hemi/electrumx/conn.go
+++ b/hemi/electrumx/conn.go
@@ -25,6 +25,7 @@ const (
 // clientConn is a connection with an ElectrumX server.
 type clientConn struct {
 	mx sync.Mutex
+	wg sync.WaitGroup
 
 	conn      net.Conn
 	requestID uint64
@@ -40,6 +41,8 @@ func newClientConn(conn net.Conn, onClose func(c *clientConn)) *clientConn {
 		closeCh: make(chan struct{}),
 		onClose: onClose,
 	}
+
+	c.wg.Add(1)
 	go c.pinger()
 	return c
 }
@@ -152,6 +155,8 @@ func (c *clientConn) ping() error {
 
 // pinger pings each connection on a ticker.
 func (c *clientConn) pinger() {
+	defer c.wg.Done()
+
 	ticker := time.NewTicker(connPingInterval)
 	for {
 		select {
@@ -185,5 +190,12 @@ func (c *clientConn) Close() error {
 		close(c.closeCh)
 		c.closeCh = nil
 	}
-	return c.conn.Close()
+
+	if err := c.conn.Close(); err != nil {
+		return err
+	}
+
+	// Wait for pinger to exit.
+	c.wg.Wait()
+	return nil
 }

--- a/hemi/electrumx/conn.go
+++ b/hemi/electrumx/conn.go
@@ -193,10 +193,10 @@ func (c *clientConn) Close() error {
 
 	if c.closeCh != nil {
 		close(c.closeCh)
-		c.closeCh = nil
 	}
 
 	// Wait for pinger to exit.
 	c.wg.Wait()
+	c.closeCh = nil
 	return nil
 }

--- a/hemi/electrumx/conn.go
+++ b/hemi/electrumx/conn.go
@@ -181,6 +181,9 @@ func (c *clientConn) Close() error {
 	if c.onClose != nil {
 		c.onClose(c)
 	}
-	close(c.closeCh)
+	if c.closeCh != nil {
+		close(c.closeCh)
+		c.closeCh = nil
+	}
 	return c.conn.Close()
 }

--- a/hemi/electrumx/conn_test.go
+++ b/hemi/electrumx/conn_test.go
@@ -33,9 +33,9 @@ func TestClientConn(t *testing.T) {
 	if err != nil {
 		t.Fatalf("failed to dial server: %v", err)
 	}
-	defer conn.Close()
 
 	c := newClientConn(conn, nil)
+	defer c.Close()
 
 	tests := []struct {
 		name   string
@@ -162,6 +162,33 @@ func TestReadResponse(t *testing.T) {
 					string(p), string(want))
 			}
 		})
+	}
+}
+
+func TestClose(t *testing.T) {
+	server := createMockServer(t)
+	defer server.Close()
+
+	conn, err := net.Dial("tcp", server.address)
+	if err != nil {
+		t.Fatalf("failed to dial server: %v", err)
+	}
+
+	c := newClientConn(conn, nil)
+
+	// Ping the server.
+	if err := c.ping(); err != nil {
+		t.Errorf("failed to ping server: %v", err)
+	}
+
+	// Close the client.
+	if err := c.Close(); err != nil {
+		t.Errorf("failed to close client: %v", err)
+	}
+
+	// Close the client again. This should do nothing and return net.ErrClosed.
+	if err := c.Close(); err == nil || !errors.Is(err, net.ErrClosed) {
+		t.Errorf("failed to close client (second): %v", err)
 	}
 }
 

